### PR TITLE
Interop with .NET `Stream` type

### PR DESCRIPTION
**Pull request by AaronRobinsonMSFT**
## Summary

This is a sample for working with the .NET [`Stream`](https://learn.microsoft.com/dotnet/api/system.io.stream) type via C exports or through COM's [`IStream`](https://learn.microsoft.com/windows/win32/api/objidl/nn-objidl-istream) interface.

### CODE REVIEW

**Comment by AaronRobinsonMSFT** [comment]
@BillWagner Can you help explain the issue here?

File: core/interop/Streams/ManagedLib/StreamAsIStream.cs [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+    {
+        public MemoryStreamWrapper(MemoryStream s)
+        {
+            this.Stream = s;
+        }
+
+        public MemoryStream Stream { get; private set; }
+
+        // ISequentialStream
+        public void Read([MarshalAs(UnmanagedType.LPArray, SizeParamIndex = 1), Out] byte[] pv, int cb, IntPtr pcbRead)
```
Comment (jkotas) [inline core/interop/Streams/ManagedLib/StreamAsIStream.cs:90]: Should this rather take `byte* pv`? As written, the unmanaged -> managed marshalling has to create an extra copy of the buffer as `byte[]` that is quite inefficient.

Comment (AaronRobinsonMSFT) [reply core/interop/Streams/ManagedLib/StreamAsIStream.cs:90]: Good suggestion. I can do that.

File: core/interop/Streams/ManagedLib/StreamAsIStream.cs [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+        catch (Exception e)
+        {
+            return e.HResult;
+        }
+        return 0;
+    }
+
+#if NETFRAMEWORK
+    [ComVisible(true)]
+    sealed partial class MemoryStreamWrapper : IStream
```
[the remark this answers is not in the document]
Comment (AaronRobinsonMSFT) [reply core/interop/Streams/ManagedLib/StreamAsIStream.cs]: Nope, converted to `MemoryStream`.

**Comment by gewarren** [comment]
@adegeo Can you help with the Snippets 5000 error?

File: core/interop/Streams/ManagedLib/StreamAsCExports.cs [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+            GCHandle.FromIntPtr(streamMaybe).Free();
+        }
+    }
+
+#if NETFRAMEWORK
+    [DNNE.Export]
+#else
+    [UnmanagedCallersOnly]
+#endif // !NETFRAMEWORK
+    public static int C_PrintStream(IntPtr streamMaybe)
```
Comment (stephentoub) [inline core/interop/Streams/ManagedLib/StreamAsCExports.cs:40]: "streamMaybe"?  The code as written will throw and then fail with an HResult if it's not.

File: core/interop/Streams/ManagedLib/StreamAsCExports.cs [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+        try
+        {
+            var stream = (Stream)GCHandle.FromIntPtr(streamMaybe).Target!;
+            Span<byte> data = new(dataRaw, (int)length);
+#if NETFRAMEWORK
+            var buffer = new byte[data.Length];
+            stream.Read(buffer, 0, buffer.Length);
+            buffer.CopyTo(data);
+#else
+            stream.Read(data);
```
Comment (stephentoub) [inline core/interop/Streams/ManagedLib/StreamAsCExports.cs:82]: These Read calls are not guaranteed to read `length` bytes.  And the shape of this C_Stream_Read method means the caller doesn't know how many bytes were actually read.

File: core/interop/Streams/ManagedLib/StreamAsCExports.cs [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+    }
+
+#if NETFRAMEWORK
+    [DNNE.Export]
+#else
+    [UnmanagedCallersOnly]
+#endif // !NETFRAMEWORK
+    public static int C_Stream_Write(IntPtr streamMaybe, nint length, byte* dataRaw)
+    {
+        Debug.Assert(length < int.MaxValue);
```
Comment (stephentoub) [inline core/interop/Streams/ManagedLib/StreamAsCExports.cs:99]: `<`, or `<=`?

**Comment by AaronRobinsonMSFT** [comment]
Ping @adegeo or @BillWagner

**Review by gewarren — APPROVED** [review]

**Comment by AaronRobinsonMSFT** [comment]
@dotnet/interop-contrib CI is all green.

File: core/interop/Streams/readme.md [hunk 6]
```diff
@@ -0,0 +1,37 @@
+# Interop with .NET Stream sample
+
+The [`Stream`](https://learn.microsoft.com/dotnet/api/system.io.stream) data type is one of the most common non-trivial data types for interop. This sample provides two distinct mechanism for working with the `Stream` type in interop scenarios.
+
+1) The projection of the `Stream` type using C-style exports.
+
+1) The projection of the `Stream` type using the COM [`IStream`](https://learn.microsoft.com/windows/win32/api/objidl/nn-objidl-istream) interface. This approach utilizes [`ComWrappers`](https://learn.microsoft.com/dotnet/api/system.runtime.interopservices.comwrappers) and the COM source generator.
```
Comment (elinor-fung) [inline core/interop/Streams/readme.md:7]: I know we don't have COM source generator docs yet - do we want to just link to the GeneratedComInterface API doc for now

File: core/interop/Streams/Shared/IStream.cs [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+            mtime = unmanaged.mtime,
+            ctime = unmanaged.ctime,
+            atime = unmanaged.atime,
+            grfMode = unmanaged.grfMode,
+            grfLocksSupported = unmanaged.grfLocksSupported,
+            clsid = unmanaged.clsid,
+            grfStateBits = unmanaged.grfStateBits,
+            reserved = unmanaged.reserved,
+        };
+
```
Comment (elinor-fung) [inline core/interop/Streams/Shared/IStream.cs:88]:
```suggestion
```

**Review by elinor-fung — APPROVED** [review]

**Pull request merged**